src/GitHub.Api/Installer/GitInstaller.cs

```diff
@@ -116,45 +117,84 @@ public void SetupGitIfNeeded(ActionTask<NPath> onSuccess, ITask onFailure)
                 return;
             }
 
-            new ActionTask(cancellationToken, () => {
-                if (IsGitExtracted())
+            var isGitExtractedTask = new FuncTask<NPath>(cancellationToken, () =>
+            {
+                if (!IsGitExtracted())
                 {
-                    Logger.Trace("SetupGitIfNeeded: Skipped");
-                    onSuccess.PreviousResult = installDetails.GitExecutablePath;
-                    onSuccess.Start();
+                    GrabZipFromResources();
+                    return null;
                 }
-                else
+                Logger.Trace("SetupGitIfNeeded: Skipped");
+                return installDetails.GitExecutablePath;
+            });
+            isGitExtractedTask.OnEnd += (t, res, _, __) =>
+            {
+                if (res == null)
                 {
-                    ExtractPortableGit(onSuccess, onFailure);
+                    var extractTask = ExtractPortableGit();
+                    extractTask.Then(onSuccess, TaskRunOptions.OnSuccess, taskIsTopOfChain: true);
+                    extractTask.Then(onFailure, TaskRunOptions.OnFailure, taskIsTopOfChain: true);
+                    t.Then(extractTask);
                 }
-            }).Start();
+                else
+                    t.Then(onSuccess);
+            };
+
+            isGitExtractedTask.Start();
         }
 
-        private void ExtractPortableGit(ActionTask<NPath> onSuccess, ITask onFailure)
+        private void GrabZipFromResources()
         {
-            ITask downloadFilesTask = null;
-            if ((gitArchiveFilePath == null) || (gitLfsArchivePath == null))
-            {
-                downloadFilesTask = CreateDownloadTask();
-            }
+            if (gitArchiveFilePath == null || !gitArchiveFilePath.FileExists())
+                gitArchiveFilePath = AssemblyResources.ToFile(ResourceType.Platform, "git.zip", installDetails.ZipPath, environment);
+            if (!gitArchiveFilePath.FileExists())
+                gitArchiveFilePath = null;
+
+            if (gitLfsArchivePath == null || !gitLfsArchivePath.FileExists())
+                gitLfsArchivePath = AssemblyResources.ToFile(ResourceType.Platform, "git-lfs.zip", installDetails.ZipPath, environment);
+            if (!gitLfsArchivePath.FileExists())
+                gitLfsArchivePath = null;
+        }
+
+        private ITask CreateDownloadTask()
+        {
+            gitArchiveFilePath = installDetails.ZipPath.Combine("git.zip");
+            gitLfsArchivePath = installDetails.ZipPath.Combine("git-lfs.zip");
 
+            var downloader = new Downloader();
+            downloader.QueueDownload(installDetails.GitZipUrl, installDetails.GitZipMd5Url, installDetails.ZipPath);
+            downloader.QueueDownload(installDetails.GitLfsZipUrl, installDetails.GitLfsZipMd5Url, installDetails.ZipPath);
```

Comment:
Do you want to download both if one or the other isn't found?

Reply:
Good point, probably not.